File: rpi_send_alert.py
#!/usr/bin/env python3
"""
🚨 Alert Sender for Raspberry Pi
Sends alerts to Railway Alert API
"""

import requests
import json
import time
from datetime import datetime

# Configuration - Update these with your values
RAILWAY_API_URL = "https://web-production-07eda.up.railway.app/api/alerts"  # Your Railway URL
DEVICE_ID = "3d49c55d-bbfd-4bd0-9663-8728d64743ac"  # Your Raspberry Pi device ID (CORRECTED)
DEVICE_NAME = "raspberrypi"

def send_alert(risk_level="Medium", description="Test alert from Raspberry Pi"):
    """Send an alert to the Railway API"""
    
    alert_payload = {
        "deviceId": DEVICE_ID,
        "alert": {
            "notification_type": "Alert",
            "detected_objects": ["test", "detection"],
            "risk_label": risk_level,
            "predicted_risk": risk_level,
            "description": [description, f"Sent from {DEVICE_NAME}"],
            "screenshot": [],
            "device_identifier": DEVICE_NAME,
            "timestamp": int(time.time() * 1000),
            "model_version": "v1.0",
            "confidence_score": 0.85,
            "additional_data": {
                "test": True,
                "source": "raspberry_pi",
                "sent_at": datetime.now().isoformat()
            }
        }
    }
    
    try:
        print(f"🚨 Sending {risk_level} alert...")
        print(f"📡 API URL: {RAILWAY_API_URL}")
        
        response = requests.post(
            RAILWAY_API_URL,
            json=alert_payload,
            headers={"Content-Type": "application/json"},
            timeout=10
        )
        
        print(f"✅ Response Status: {response.status_code}")
        
        if response.status_code == 200:
            result = response.json()
            print(f"✅ Alert sent successfully!")
            print(f"📋 Alert ID: {result.get('alertId')}")
            print(f"📱 Notification: {result.get('notification', {}).get('title')}")
            return True
        else:
            print(f"❌ Error: {response.text}")
            return False
            
    except Exception as error:
        print(f"❌ Error sending alert: {error}")
        return False

def test_health():
    """Test if the API is healthy"""
    try:
        health_url = RAILWAY_API_URL.replace("/api/alerts", "/health")
        print(f"🏥 Testing health endpoint...")
        response = requests.get(health_url, timeout=5)
        
        if response.status_code == 200:
            data = response.json()
            print(f"✅ API is healthy")
            print(f"   Firebase: {'✅ Connected' if data.get('firebase') else '⚠️  Disabled'}")
            return True
        else:
            print(f"❌ Health check failed: {response.status_code}")
            return False
    except Exception as error:
        print(f"❌ Health check failed: {error}")
        return False

def main():
    print("=" * 60)
    print("🚨 Raspberry Pi Alert Sender")
    print("=" * 60)
    print(f"Device: {DEVICE_NAME}")
    print(f"Device ID: {DEVICE_ID}")
    print(f"API URL: {RAILWAY_API_URL}")
    print("=" * 60)
    print()
    
    # Test health first
    if not test_health():
        print("\n❌ API is not responding. Check your Railway URL and internet connection.")
        return False
    
    print("\n📤 Sending test alerts...")
    print("-" * 60)
    
    # Send different risk level alerts
    alerts = [
        ("Low", "Low risk detection test"),
        ("Medium", "Medium risk detection test"),
        ("High", "High risk detection test"),
        ("Critical", "Critical risk detection test"),
    ]
    
    for risk_level, description in alerts:
        send_alert(risk_level, description)
        print()
        time.sleep(2)  # Wait 2 seconds between alerts
    
    print("=" * 60)
    print("✅ Test complete!")
    print("📱 Check your mobile app for notifications")
    print("=" * 60)
    return True

if __name__ == "__main__":
    main()

File: rpi_send_alert.js
#!/usr/bin/env node

/**
 * 🚨 Alert Sender for Raspberry Pi (Node.js Version)
 * Sends alerts to Alert API and checks access via Admin Portal
 */

const fs = require('fs');
const path = require('path');
const http = require('http');
const https = require('https');

require('dotenv').config();

// Configuration
const EC2_HOST = process.env.EC2_HOST || '13.205.201.82';

const ALERT_API_URL =
  process.env.ALERT_API_URL ||
  process.env.RAILWAY_API_URL ||
  `http://${EC2_HOST}/alert-api/api/alerts`;

const ADMIN_PORTAL_URL = process.env.ADMIN_PORTAL_URL || `http://${EC2_HOST}`;
const API_KEY = process.env.API_KEY || process.env.ADMIN_API_KEY || '';
const CHECK_ACCESS_BEFORE_SEND = (process.env.CHECK_ACCESS_BEFORE_SEND || 'false').toLowerCase() === 'true';
const configuredDeviceIdFile = process.env.DEVICE_ID_FILE || './device_id.txt';
const DEVICE_ID_FILE = path.isAbsolute(configuredDeviceIdFile)
  ? configuredDeviceIdFile
  : path.resolve(process.cwd(), configuredDeviceIdFile);
const DEVICE_NAME = "raspberrypi";

/**
 * Read device ID from file
 */
function readDeviceId() {
  try {
    if (!fs.existsSync(DEVICE_ID_FILE)) {
      console.error(`❌ Error: device ID file not found at ${DEVICE_ID_FILE}`);
      process.exit(1);
    }
    
    const deviceId = fs.readFileSync(DEVICE_ID_FILE, 'utf8').trim();
    
    if (!deviceId) {
      console.error('❌ Error: device ID file is empty');
      process.exit(1);
    }
    
    return deviceId;
  } catch (error) {
    console.error('❌ Error reading device ID file:', error.message);
    process.exit(1);
  }
}

/**
 * Send HTTP request (GET or POST)
 */
function sendRequest(url, data, method = 'POST', extraHeaders = {}) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const isHttps = urlObj.protocol === 'https:';
    const client = isHttps ? https : http;
    
    const options = {
      hostname: urlObj.hostname,
      port: urlObj.port || (isHttps ? 443 : 80),
      path: urlObj.pathname + urlObj.search,
      method: method,
      headers: {
        'Content-Type': 'application/json'
      },
      timeout: 5000,
    };

    options.headers = { ...options.headers, ...extraHeaders };

    if (method === 'POST' && data) {
      options.headers['Content-Length'] = Buffer.byteLength(JSON.stringify(data));
    }

    const req = client.request(options, (res) => {
      let responseData = '';

      res.on('data', (chunk) => {
        responseData += chunk;
      });

      res.on('end', () => {
        try {
          const parsed = JSON.parse(responseData);
          resolve({ status: res.statusCode, data: parsed });
        } catch (e) {
          resolve({ status: res.statusCode, data: responseData });
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.on('timeout', () => {
      req.destroy(new Error('Request timeout'));
    });

    if (method === 'POST' && data) {
      req.write(JSON.stringify(data));
    }
    req.end();
  });
}

/**
 * Check whether the device is allowed to send alerts via Admin Portal.
 */
async function checkDeviceStatus(deviceId) {
  const url = `${ADMIN_PORTAL_URL}/api/check-device/${encodeURIComponent(deviceId)}`;

  try {
    const response = await sendRequest(url, null, 'GET', {
      'X-API-Key': API_KEY,
    });

    if (response.status !== 200 || typeof response.data !== 'object') {
      return {
        reachable: false,
        hasAccess: true,
        reason: `Unexpected admin portal response: HTTP ${response.status}`,
      };
    }

    return {
      reachable: true,
      hasAccess: response.data.hasAccess !== false,
      registered: response.data.registered !== false,
      reason: response.data.reason || null,
      device: response.data.device || null,
    };
  } catch (error) {
    return {
      reachable: false,
      hasAccess: true,
      reason: `Admin portal device check failed: ${error.message}`,
    };
  }
}

/**
 * Send an alert to the Alert API
 */
async function sendAlert(deviceId, riskLevel = "Medium", description = "Test alert from Raspberry Pi") {
  const alertPayload = {
    deviceId: deviceId,
    alert: {
      notification_type: "Alert",
      detected_objects: ["test", "detection"],
      risk_label: riskLevel,
      predicted_risk: riskLevel,
      description: [description, `Sent from ${DEVICE_NAME}`],
      screenshot: [],
      device_identifier: DEVICE_NAME,
      timestamp: Date.now(),
      model_version: "v1.0",
      confidence_score: 0.85,
      additional_data: {
        test: true,
        source: "raspberry_pi",
        sent_at: new Date().toISOString()
      }
    }
  };

  try {
    console.log(`🚨 Sending ${riskLevel} alert...`);
    console.log(`📡 API URL: ${ALERT_API_URL}`);

    const response = await sendRequest(ALERT_API_URL, alertPayload);

    console.log(`✅ Response Status: ${response.status}`);

    if (response.status === 200) {
      console.log(`✅ Alert sent successfully!`);
      console.log(`📋 Alert IDs: ${JSON.stringify(response.data.alertIds)}`);
      console.log(`👥 Users notified: ${response.data.usersNotified}`);
      return true;
    } else {
      console.log(`❌ Error: ${JSON.stringify(response.data)}`);
      return false;
    }
  } catch (error) {
    console.error(`❌ Error sending alert: ${error.message}`);
    return false;
  }
}

/**
 * Test if the API is healthy
 */
async function testHealth() {
  const healthUrl = ALERT_API_URL.replace("/api/alerts", "/health");
  
  try {
    console.log(`🏥 Testing health endpoint...`);
    const response = await sendRequest(healthUrl, null, 'GET');

    if (response.status === 200) {
      console.log(`✅ API is healthy`);
      console.log(`   Firebase: ${response.data.firebase ? '✅ Connected' : '⚠️  Disabled'}`);
      return true;
    } else {
      console.log(`❌ Health check failed: ${response.status}`);
      return false;
    }
  } catch (error) {
    console.error(`❌ Health check failed: ${error.message}`);
    return false;
  }
}

/**
 * Main function
 */
async function main() {
  const deviceId = readDeviceId();
  const shouldRunAccessPrecheck = CHECK_ACCESS_BEFORE_SEND && !!API_KEY;

  console.log("=".repeat(60));
  console.log("🚨 Raspberry Pi Alert Sender (Node.js)");
  console.log("=".repeat(60));
  console.log(`Device: ${DEVICE_NAME}`);
  console.log(`Device ID: ${deviceId}`);
  console.log(`Device ID File: ${DEVICE_ID_FILE}`);
  console.log(`Alert API URL: ${ALERT_API_URL}`);
  console.log(`Admin Portal URL: ${ADMIN_PORTAL_URL}`);
  console.log(`Device precheck: ${shouldRunAccessPrecheck ? 'enabled' : 'disabled'}`);
  console.log("=".repeat(60));
  console.log();

  if (CHECK_ACCESS_BEFORE_SEND && !API_KEY) {
    console.warn('⚠️  CHECK_ACCESS_BEFORE_SEND is enabled, but API_KEY/ADMIN_API_KEY is not set.');
    console.warn('⚠️  Skipping device precheck; alert API will enforce final device restrictions.');
    console.log();
  }

  if (shouldRunAccessPrecheck) {
    console.log('🔐 Checking device status via admin portal...');
    const access = await checkDeviceStatus(deviceId);

    if (!access.reachable) {
      console.warn(`⚠️  ${access.reason}`);
      console.warn('⚠️  Continuing in fail-open mode; alert API will do final device checks.');
    } else if (!access.hasAccess) {
      console.error('🚫 Device is not allowed to send alerts.');
      console.error(`   Reason: ${access.reason || 'No access'}`);
      process.exit(1);
    } else {
      console.log('✅ Device is allowed to send alerts');
    }

    console.log();
  }

  // Test health first
  if (!(await testHealth())) {
    console.log("\n❌ Alert API is not responding. Check ALERT_API_URL and internet connection.");
    process.exit(1);
  }

  console.log("\n📤 Sending test alerts...");
  console.log("-".repeat(60));

  // Send different risk level alerts
  const alerts = [
    ["Low", "Low risk detection test"],
    ["Medium", "Medium risk detection test"],
    ["High", "High risk detection test"],
    ["Critical", "Critical risk detection test"]
  ];

  for (const [riskLevel, description] of alerts) {
    await sendAlert(deviceId, riskLevel, description);
    console.log();
    
    // Wait 2 seconds between alerts
    await new Promise(resolve => setTimeout(resolve, 2000));
  }

  console.log("=".repeat(60));
  console.log("✅ Test complete!");
  console.log("📱 Check your mobile app for notifications");
  console.log("=".repeat(60));
}

// Run main function
main().catch(error => {
  console.error("❌ Fatal error:", error);
  process.exit(1);
});
